File: app/src/main/java/me/instabattle/app/managers/PhotoManager.java
package me.instabattle.app.managers;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.cloudinary.Cloudinary;
import com.cloudinary.Transformation;
import com.cloudinary.utils.ObjectUtils;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import me.instabattle.app.images.Util;

public class PhotoManager {
    private static final String TAG = "PhotoManager";
    private static final Cloudinary cloudinary;
    private static final Map<String, String> config = new HashMap<>();

    static {
        config.put("cloud_name", "instabattle");
        config.put("api_key", "277887798658767");
        config.put("api_secret", "V3VsI9jHpatwH1nsVgf6NP11nSg");
        cloudinary = new Cloudinary(config);
    }

    public static void getPhotoAndDo(String name, final BitmapCallback callback) {
        final Exception exception;
        try {
            final URL url = new URL(cloudinary.url().generate(name) + ".jpg");
            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        final Bitmap image = Util.decodeSampledBitmapFromURL(url, 256, 256);
                        callback.onResponse(image);
                        Log.d(TAG, "Got photo from " + url.toString());
                    } catch (IOException e) {
                        Log.e(TAG, "cannot establish connection with " + url.toString());
                        callback.onFailure(e);
                    }
                }
            }).start();
            return;
        } catch (MalformedURLException e) {
            Log.e(TAG, "can't create url from name");
            exception = e;
        }
        new Thread(() -> callback.onFailure(exception)).start();
    }

    public static void upload(final String name, final byte[] photo) {
        new Thread(() -> {
            try {
                Map params = ObjectUtils.asMap("public_id", name);
                cloudinary.uploader().upload(new ByteArrayInputStream(photo), params);
            } catch (IOException e) {
                Log.e("Cloudinary fail", "can't upload photo");
            }
        }).start();
    }
}

File: app/src/main/kotlin/me/instabattle/app/managers/PhotoManager.kt
package me.instabattle.app.managers

import android.content.Context
import android.widget.ImageView
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions

import com.cloudinary.Cloudinary
import com.cloudinary.utils.ObjectUtils
import me.instabattle.app.GlideApp

import java.io.ByteArrayInputStream
import java.io.IOException

import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.error
import kotlin.concurrent.thread

object PhotoManager: AnkoLogger {
    private val config = mapOf(
            "cloud_name" to "instabattle",
            "api_key" to "277887798658767",
            "api_secret" to "V3VsI9jHpatwH1nsVgf6NP11nSg"
    )
    private val cloudinary = Cloudinary(config)

    fun getPhotoInto(ctx: Context, name: String, imageView: ImageView) {
        val url = cloudinary.url().generate(name) + ".jpg"
        //val url = URL(cloudinary.url().generate(name) + ".jpg")
        //val image = Util.decodeSampledBitmapFromURL(url, 256, 256)
        GlideApp
                .with(ctx)
                .load(url)
                //.placeholder(R.mipmap.ic_launcher) // can also be a drawable
                //.error(R.mipmap.future_studio_launcher) // will be displayed if the image cannot be loaded
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(imageView);
    }

    fun upload(name: String, photo: ByteArray) {
        thread {
            try {
                val params = ObjectUtils.asMap("public_id", name)
                cloudinary.uploader().upload(ByteArrayInputStream(photo), params)
            } catch (e: IOException) {
                error("Can't upload photo")
            }
        }
    }
}
